tools/nl/embeddings/finetune.py

```diff
@@ -0,0 +1,258 @@
+  """Fine tuning involves providing pairs of sentences/text with an
+  approximate similar score to a baseline model. These pairs (and scores)
+  are used for further `training` (with CosineSimilarityLoss as the objecive).
+  The end result is that the baseline model's weights get updated based on
+  the `new` training examples (pairs).
+  Effectively, the training examples (pairs and scores) provide additional
+  context to a baseline model about the kinds of associations between
+  sentences/text that we care about. In this case, we use the StatVar name,
+  description and alternatives (human curated and LLM-generated) to create
+  pairs of sentences/text with high similarity scores. We also provide a
+  way to indicate that some pairs of texts/words/sentences should not be
+  associated with each other by assigning very low scores. This allows us to
+  bias the model's weights to accomodate our particular use case better. 
+  """
+
+  training_examples: List[InputExample] = []
+
+  # Iterate over SV Names, Descriptions and Alternatives to produce
+  # pairs of texts which we want to associated with each other in
+  # terms of similarity.
+  for _, row in df_svs.iterrows():
+    name = row[utils.NAME_COL]
+    descriptions = row[utils.DESCRIPTION_COL].split(";")
+    desc = descriptions[0]
+
+    if not name and not desc:
+      continue
+    if ((not name) and desc):
+      name = desc
+    elif ((not desc) and name):
+      desc = name
+
+    curated = row[utils.CURATED_ALTERNATIVES_COL].split(";")
+    palm_alts = row[utils.ALTERNATIVES_COL].split(";")
+
+    # Pair name and description with very high score.
+    if name != desc:
+      training_examples.append(
+          InputExample(texts=[name, desc], label=VERY_HIGH_MATCH_SCORE))
+
+    # If there are more descriptions, pair them with the name as well.
+    if len(descriptions) > 1:
+      for i in range(1, len(descriptions)):
+        if descriptions[i]:
+          training_examples.append(
+              InputExample(texts=[name, descriptions[i]],
+                           label=HIGH_MATCH_SCORE))
+
+    for c in curated:
+      # All all curated alternatives as pairs with the description.
+      if c:
+        training_examples.append(
+            InputExample(texts=[desc, c], label=HIGH_MATCH_SCORE))
+
+    for p in palm_alts:
+      if p:
+        # High match score but since these are auto-generated, keep the score lower
+        # than for the manual/curated cases.
+        training_examples.append(
+            InputExample(texts=[name, p], label=MEDIUM_HIGH_MATCH_SCORE))
+
+  # Now add the manual sentence pairs (which are assumed to have the scores).
+  for _, row in df_sentence_pairs.iterrows():
+    assert "sentence_1" in row
+    assert "sentence_2" in row
+    assert "score" in row
+    training_examples.append(
+        InputExample(texts=[row["sentence_1"], row["sentence_2"]],
+                     label=row["score"]))
```

Comment:
nit: We can have a `assert float(row["score"])`.

Reply:
Done